### PR TITLE
KAFKA-2944: Replaced the NPE with a nicer error and clean exit and added debug me…

### PR DESCRIPTION
**Pull request by gwenshap**
…ssage to assist with figuring this out.

### CODE REVIEW

**Comment by gwenshap** [comment]
Note: This doesn't exactly resolve KAFKA-2944 (i.e. we may still get into a bad state with commits we can't account for), but since it does change the error from NPE to "Received a commit message for connector " + connectorName + " but there is no matching configuration for tasks in this connector. This should never happen.", with less extreme results (i.e. the log reading thread should not blow up), I think this may allow up to close this specific JIRA and continue with a different one debugging a less severe condition.

**Comment by gwenshap** [comment]
Another note: This adds a debug log for every single config read and write. I'm assuming configs are usually not huge and we are not re-configuring that frequently. If the logs will become unreadable, we can revisit.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigStorage.java [hunk 1]
```diff
@@ -396,6 +398,7 @@ public void onCompletion(Throwable error, ConsumerRecord<String, byte[]> record)
                 synchronized (lock) {
                     if (value.value() == null) {
                         // Connector deletion will be written as a null value
+                        log.info("Removed connector " + connectorName + " due to null configuration. This is usually intentional and does not indicate an issue.");
```
Comment (ewencp) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigStorage.java:402]: Do you actually want this at info level? This looks like the one case that will add log messages at the default level. Deletions ideally aren't all that frequent, but on worker start up we'll log all the deletions on INFO level as the config log is consumed to catch up.

**Comment by ewencp** [comment]
@gwenshap Left one comment and also think you may have missed logging something in `putConnectorConfig`. Otherwise LGTM and am happy to commit after answering these questions.

I suppose one last thing is if we want to file that follow-up JIRA, we could put the JIRA ID either into the code or the error message so they both get resolved together when we finally track down the issue.

File: connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigStorage.java (hunk 1 above)
Comment (gwenshap) [inline connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigStorage.java:402]: Deletions will eventually be compacted away, so I figured it should not be too painful. 
IMO a connector completely going away is a big event that someone watching logs will want to know about - which is why I chose INFO. Maybe I am underestimating the amount of log crap that will show up on restart - I imagined that each connector will have 5-10 connectors, so we are talking about ~ 10 messages at startup. Do you foresee different numbers? Or you think the 10 extra messages are already too much?

**Comment by gwenshap** [comment]
added the missing log and reference to a JIRA. IMO the "info" log should stay, unless I'm underestimating how annoying it will be.

**Comment by ewencp** [comment]
Explanation on INFO level makes sense.

LGTM

**Pull request closed without merging**